pulp_2to3_migration/app/models/base.py

```diff
@@ -11,24 +19,136 @@ class MigrationPlan(Model):
     """
     plan = JSONField()
+    _real_plan = None
+
+    @property
+    def plan_view(self):
+        """
+        Window to view the validated migration plan data through.
+        """
+        if not self._real_plan:
+            self._real_plan = _InternalMigrationPlan(self.plan)
+
+        return self._real_plan
+
+    def get_plugins(self):
+        """
+        Return a list of pulp2 plugins to migrate.
+        """
+        return self.plan_view.plugins_to_migrate
 
     def get_repositories(self):
         """
         Return a list of pulp2 repositories to migrate or empty list if all should be migrated.
         """
-        # TODO: get pulp2 repositories from the self.plan
-        return []
+        return self.plan_view.repositories_to_migrate
 
     def get_importers(self):
         """
         Return a list of pulp2 importers to migrate or empty list if all should be migrated.
         """
-        # TODO: get pulp2 importers from the self.plan
-        return []
+        return self.plan_view.importers_to_migrate
 
     def get_distributors(self):
         """
         Return a list of pulp2 distributors to migrate or empty list if all should be migrated.
         """
-        # TODO: get pulp2 distributors from the self.plan
-        return []
+        return self.plan_view.distributors_to_migrate
+
+
+class PlanValidationFailure(Exception):
+    """
+    Exception to be thrown when validating the MigrationPlan.
+
+    e.g. Repository specified does not exist.
+    """
+    pass
+
+
+class _InternalMigrationPlan:
+    def __init__(self, migration_plan):
+        self.migration_plan = migration_plan
+
+        self.plugins_to_migrate = []
+        self.importers_to_migrate = []
+        self.distributors_to_migrate = []
+
+        self.importer_discovery_repos = []
+
+        self.repositories_to_create = []
+
+        self.missing_importers = []
+        self.missing_repositories = []
+        self.missing_distributors = []
+
+        # Make sure we've initialized the MongoDB connection first
+        connection.initialize()
+        self.populate()
+
+        self.validated = False
+
+    def populate(self):
+        for plugin_data in self.migration_plan['plugins']:
+            self.plugins_to_migrate.append(plugin_data['type'])
+            if plugin_data.get('repositories'):
+                self._parse_repository_data(plugin_data.get('repositories'))
+                # TODO: do something with protection
+
+    def _parse_repository_data(self, repository_data):
+        for repository in repository_data:
+            name = repository_data['name']
+
+            _find_importer_repo = repository['pulp2_repository_id']
+            self.importer_discovery_repos.append(_find_importer_repo)
+
+            repository_versions = [] or \
+                self._parse_repository_version_data(repository.get('repository_versions'))
+
+            repository_data = {
```

Comment:
i think then intention was to include _find_importer_repo inside of this structure. This way we will know which importer to use when creating remote for repo named X form this repo_data structure. did you have something else on your mind?

Reply:
+1 to add importer info, we need to relate it later. After migrating everyting to Pulp3, user needs to know which Remote to use with which Repository. And this information is available here only.

It would also be convenient to have a dictionary here instead of a list, because repositories are created first and then they are getting migrated so getting information about repo versions by `name` would be nice.